pylabrobot/heating_shaking/hamilton.py

```diff
@@ -0,0 +1,81 @@
+from pylabrobot.heating_shaking.backend import HeaterShakerBackend
+from pylabrobot.machines.backends.usb import USBBackend
+from enum import Enum
+
+class PlateLockPosition(Enum):
+  LOCKED = 1
+  UNLOCKED = 0
+
+class Command(Enum):
+  INITIALIZE_PLATE_LOCK = "LI"
+  MOVE_PLATE_LOCK = "LP"
+  SHAKE = "SB"
+  STOP_SHAKING = "SC"
+  WAIT_FOR_STOP = "SW"
+
+class RotationDirection(Enum):
+  CLOCKWISE = 0
+  COUNTER_CLOCKWISE = 1
+
+class HamiltonHeatShaker(HeaterShakerBackend, USBBackend):
+  """
+  Backend for Hamilton Heater Shaker devices connected through
+  an HSB
+  """
+
+  def __init__(self, shaker_index:int, id_vendor:int = 2223, id_product:int = 32770) -> None:
+    """
+    Multiple Hamilton Heater Shakers can be connected to the same HSB. Each has a separate 'shaker index'
+    """
+    assert shaker_index >= 0, "Shaker index must be non-negative"
+    self.shaker_index = shaker_index
+    self.command_id = 0
+
+    HeaterShakerBackend.__init__(self)
+    USBBackend.__init__(self, id_vendor, id_product)
+
+  async def setup(self):
+    await USBBackend.setup(self)
+
+  async def stop(self):
+    await USBBackend.stop(self)
+
+  def _send_command(self, command: Command, **kwargs):
+    args = "".join([f"{key}{value}" for key, value in kwargs.items()])
+    USBBackend.write(self, f"T{self.shaker_index}{command.value}id{str(self.command_id).zfill(4)}{args}")
+
+    self.command_id = (self.command_id + 1) % 10_000
+    return USBBackend.read(self)
```

Comment:
Yea, I guess the idea being that all firmware commands to Hamilton follow a similar format?

1. This could or could not be true. The "rule of three" suggests we wait for a third hamilton component before doing this.
2. If we did do this, It seems like one path would be to subclass `USBBackend` to `HamiltonUSBBackend` and add the command there.
3. There are some differences already, like the `tip_pattern` param and the parse firmware string. We could overwrite these for sure, but maybe end up being more complex
4. I think the command to unify would be `assemble_command`, but I'm curious about the decision to use bools there instead of directly feeding the firmware string. Seems like the abstraction boundary isn't clearly defined. 

I tend to be on the "wait and see" side of this.

Reply:
agreed